source/MRMesh/MRExtractIsolines.cpp

```diff
@@ -127,14 +191,73 @@ std::vector<std::vector<MeshEdgePoint>> extractIsolines( const MeshTopology & to
     return s.extract();
 }
 
-std::vector<std::vector<MeshEdgePoint>> extractPlaneSections( const MeshPart & mp, const Plane3f & plane )
+PlaneSections extractPlaneSections( const MeshPart & mp, const Plane3f & plane )
 {
     MR_TIMER;
 
     Isoliner s( mp.mesh.topology, [&]( VertId v ) { return plane.distance( mp.mesh.points[v] ); }, mp.region );
     return s.extract();
 }
 
+PlaneSection trackSection( const MeshPart& mp, 
+    const MeshTriPoint& start, MeshTriPoint& end, const Vector3f& direction, float distnace )
+{
+    MR_TIMER;
+    if ( distnace == 0.0f )
+    {
+        end = start;
+        return {};
+    }
+    const auto dir = distnace > 0.0f ? direction : -direction;
+    distnace = std::abs( distnace );
+    auto startPoint = mp.mesh.triPoint( start );
+    auto prevPoint = startPoint;
+    auto plane = Plane3f::fromDirAndPt( cross( dir, mp.mesh.pseudonormal( start ) ), prevPoint );
+    ValueInVertex valueInVertex = [&] ( VertId v )
+    {
+        return plane.distance( mp.mesh.points[v] );
+    };
+    ContinueTrack continueTrack = [&] ( const MeshEdgePoint& next ) mutable->bool
+    {
+        auto point = mp.mesh.edgePoint( next );
+        auto dist = ( point - prevPoint ).length();
+        distnace -= dist;
+        if ( distnace < 0.0f )
+            return false;
+        prevPoint = point;
+        return true;
+    };
+
+    Isoliner s( mp.mesh.topology, valueInVertex, mp.region );
+    auto res = s.track( start, continueTrack );
+    if ( res.empty() )
+    {
+        end = start;
+        return {};
+    }
+    bool closed = res.size() != 1 && res.front() == res.back();
+    if ( distnace > 0.0f )
+    {
+        end = res.back();
+        res.resize( int( res.size() ) - 1 );
+        if ( closed )
+            end = start;
+        return res;
+    }
+    
```

Comment:
Can it be closed==true and distance <= 0?

Reply:
this case is handled in the last lines of the function